taichi/codegen/spirv/kernel_utils.cpp

```diff
@@ -124,6 +124,21 @@ KernelContextAttributes::KernelContextAttributes(
   TI_TRACE("rets:");
   rets_bytes_ = arange_args(&ret_attribs_vec_, 0, true, false);
 
+  TI_ASSERT(arg_attribs_vec_.size() == kernel.args_type->elements().size());
+  for (int i = 0; i < arg_attribs_vec_.size(); ++i) {
+    TI_ASSERT(arg_attribs_vec_[i].offset_in_mem ==
+              kernel.args_type->get_element_offset({i}));
+  }
+
+  TI_ASSERT(ret_attribs_vec_.size() == kernel.ret_type->elements().size());
+  for (int i = 0; i < ret_attribs_vec_.size(); ++i) {
+    TI_ASSERT(ret_attribs_vec_[i].offset_in_mem ==
+              kernel.ret_type->get_element_offset({i}));
+  }
+
+  TI_ASSERT(args_bytes_ == kernel.args_size);
```

Comment:
This may not be true? Are we ensuring that the `kernel.args_type` are constructed with the specific device capabilities of this device and have the exact same layout

Reply:
yes.